[Merged by Bors] - chore: remove more `bex` and `ball` from lemma names

grunweg opened the pull request:
Follow-up to #10816.

Remaining places containing such lemmas are
- `Option.bex_ne_none` and `Option.ball_ne_none`: defined in Lean core
- `Nat.decidableBallLT` and `Nat.decidableBallLE`: defined in Lean core
- `bef_def` is still used in a number of places and could be renamed
- `BAll.imp_{left,right}`, `BEx.imp_{left,right}`, `BEx.intro` and `BEx.elim`

I only audited the first ~150 lemmas mentioning "ball"; too many lemmas named after Metric.ball/openBall/closedBall.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

YaelDillies commented on Mathlib/Data/Set/Image.lean:
```diff
@@ -207,10 +207,11 @@ variable {f : α → β} {s t : Set α}
 #align set.image Set.image
 
-theorem mem_image_iff_bex {f : α → β} {s : Set α} {y : β} :
+theorem mem_image_iff_exists {f : α → β} {s : Set α} {y : β} :
     y ∈ f '' s ↔ ∃ (x : _) (_ : x ∈ s), f x = y :=
   bex_def.symm
```
Could you simply delete this lemma in favour of `Set.mem_image`?
```suggestion
@[deprecated mem_image] -- 2024-03-23
theorem mem_image_iff_bex {f : α → β} {s : Set α} {y : β} :
    y ∈ f '' s ↔ ∃ (x : _) (_ : x ∈ s), f x = y :=
  bex_def.symm
```

YaelDillies commented on Mathlib/Init/Data/List/Instances.lean:
```diff
@@ -62,12 +62,12 @@ variable {α : Type u} {p : α → Prop} [DecidablePred p]
 -- To work around lean4#2552, we call specific `Decidable` instances and use `match` on them,
 -- as opposed to using `if`.
-instance decidableBex : ∀ (l : List α), Decidable (∃ x ∈ l, p x)
+instance decidableExists : ∀ (l : List α), Decidable (∃ x ∈ l, p x)
```
```suggestion
instance decidableExistsMem : ∀ (l : List α), Decidable (∃ x ∈ l, p x)
```

YaelDillies commented on Mathlib/Init/Data/List/Instances.lean:
```diff
@@ -76,14 +76,17 @@ instance decidableBex : ∀ (l : List α), Decidable (∃ x ∈ l, p x)
+#align list.decidable_bex List.decidableExists
 
-instance decidableBall (l : List α) : Decidable (∀ x ∈ l, p x) :=
+instance decidableForall (l : List α) : Decidable (∀ x ∈ l, p x) :=
```
```suggestion
instance decidableForallMem (l : List α) : Decidable (∀ x ∈ l, p x) :=
```

YaelDillies commented on Mathlib/Init/Data/List/Instances.lean:
```diff
@@ -76,14 +76,17 @@ instance decidableBex : ∀ (l : List α), Decidable (∃ x ∈ l, p x)
-#align list.decidable_ball List.decidableBall
+#align list.decidable_ball List.decidableForall
+
+@[deprecated] alias List.decidableBex := List.decidableExists -- 2024-03-23
+@[deprecated] alias List.decidableBall := List.decidableForall -- 2024-03-23
```
I don't think we should be deprecating instance names.

grunweg replied:
Makes sense! Dropped.

YaelDillies commented on Mathlib/Init/Data/List/Lemmas.lean:
```diff
@@ -72,20 +72,25 @@ alias ⟨eq_or_mem_of_mem_cons, _⟩ := mem_cons
 
-theorem not_bex_nil (p : α → Prop) : ¬∃ x ∈ @nil α, p x := fun ⟨_, hx, _⟩ => List.not_mem_nil _ hx
-#align list.not_bex_nil List.not_bex_nil
+theorem not_exists_nil (p : α → Prop) : ¬∃ x ∈ @nil α, p x :=
```
```suggestion
theorem not_exists_mem_nil (p : α → Prop) : ¬∃ x ∈ @nil α, p x :=
```

grunweg replied:
This one already exists...

grunweg replied:
AFAICT, it is not used in mathlib, though.

YaelDillies commented on Mathlib/Init/Data/List/Lemmas.lean:
```diff
@@ -72,20 +72,25 @@ alias ⟨eq_or_mem_of_mem_cons, _⟩ := mem_cons
 #align list.ball_nil List.forall_mem_nil
 
-theorem bex_cons (p : α → Prop) (a : α) (l : List α) : (∃ x ∈ a :: l, p x) ↔ p a ∨ ∃ x ∈ l, p x :=
+theorem exists_cons (p : α → Prop) (a : α) (l : List α) :
```
```suggestion
theorem exists_mem_cons (p : α → Prop) (a : α) (l : List α) :
```

grunweg replied:
This one already existed, i.e. this lemmas was a duplicate. Changed the align and alias accordingly and deleted this lemma.

YaelDillies commented on Mathlib/Init/Data/List/Lemmas.lean:
```diff
@@ -72,20 +72,25 @@ alias ⟨eq_or_mem_of_mem_cons, _⟩ := mem_cons
-#align list.bex_cons List.bex_cons
+#align list.bex_cons List.exists_cons
+
+@[deprecated] alias List.not_bex_nil := List.not_exists_nil -- 2024-03-23
+@[deprecated] alias List.bex_cons := List.exists_cons -- 2024-03-23
```
```suggestion
-- 2024-03-23
@[deprecated] alias List.not_bex_nil := List.not_exists_mem_nil
@[deprecated] alias List.bex_cons := List.exists_mem_cons
```

grunweg replied:
I'm not convinced this is really better, but I don't feel strongly. Changed.

YaelDillies commented on Mathlib/Data/Set/Basic.lean:
```diff
@@ -614,9 +614,10 @@ theorem subset_eq_empty {s t : Set α} (h : t ⊆ s) (e : s = ∅) : t = ∅ :=
 #align set.subset_eq_empty Set.subset_eq_empty
 
-theorem ball_empty_iff {p : α → Prop} : (∀ x ∈ (∅ : Set α), p x) ↔ True :=
+theorem forall_empty_iff {p : α → Prop} : (∀ x ∈ (∅ : Set α), p x) ↔ True :=
```
```suggestion
theorem forall_mem_empty {p : α → Prop} : (∀ x ∈ (∅ : Set α), p x) ↔ True :=
```

YaelDillies commented on Mathlib/Data/Set/Basic.lean:
```diff
@@ -614,9 +614,10 @@ theorem subset_eq_empty {s t : Set α} (h : t ⊆ s) (e : s = ∅) : t = ∅ :=
   iff_true_intro fun _ => False.elim
-#align set.ball_empty_iff Set.ball_empty_iff
+#align set.ball_empty_iff Set.forall_empty_iff
+@[deprecated] alias Set.ball_empty_iff := Set.forall_empty_iff -- 2024-03-23
```
```suggestion
@[deprecated] alias ball_empty_iff := forall_empty_iff -- 2024-03-23
```
and below, else you're creating `Set.Set.ball_empty_iff`

YaelDillies commented on Mathlib/Data/Finset/Basic.lean:
```diff
@@ -506,9 +506,10 @@ alias ⟨_, Nonempty.to_set⟩ := coe_nonempty
 #align finset.nonempty.coe_sort Finset.Nonempty.coe_sort
 
-theorem Nonempty.bex {s : Finset α} (h : s.Nonempty) : ∃ x : α, x ∈ s :=
+theorem Nonempty.exists {s : Finset α} (h : s.Nonempty) : ∃ x : α, x ∈ s :=
```
```suggestion
theorem Nonempty.exists_mem {s : Finset α} (h : s.Nonempty) : ∃ x : α, x ∈ s :=
```

grunweg commented:
Thank you for the quick review. I have addressed all comments.

YaelDillies commented on Mathlib/Analysis/Convex/Basic.lean:
```diff
@@ -188,8 +188,8 @@ theorem convex_segment (x y : E) : Convex 𝕜 [x -[𝕜] y] := by
 
 theorem Convex.linear_image (hs : Convex 𝕜 s) (f : E →ₗ[𝕜] F) : Convex 𝕜 (f '' s) := by
   intro x hx y hy a b ha hb hab
```
```suggestion
  rintro _ ⟨x, hx, rfl⟩ _ ⟨y, hy, rfl⟩ a b ha hb hab
```

YaelDillies commented on Mathlib/Analysis/Convex/Basic.lean:
```diff
@@ -188,8 +188,8 @@ theorem convex_segment (x y : E) : Convex 𝕜 [x -[𝕜] y] := by
   intro x hx y hy a b ha hb hab
-  obtain ⟨x', hx', rfl⟩ := mem_image_iff_bex.1 hx
-  obtain ⟨y', hy', rfl⟩ := mem_image_iff_bex.1 hy
+  obtain ⟨x', hx', rfl⟩ := mem_image.1 hx
+  obtain ⟨y', hy', rfl⟩ := mem_image.1 hy
   exact ⟨a • x' + b • y', hs hx' hy' ha hb hab, by rw [f.map_add, f.map_smul, f.map_smul]⟩
```
```suggestion
  exact ⟨a • x + b • y, hs hx hy ha hb hab, by rw [f.map_add, f.map_smul, f.map_smul]⟩
```

YaelDillies commented on Mathlib/Analysis/NormedSpace/Banach.lean:
```diff
@@ -226,7 +226,7 @@ protected theorem isOpenMap (surj : Surjective f) : IsOpenMap f := by
   rcases exists_preimage_norm_le f surj with ⟨C, Cpos, hC⟩
   refine' isOpen_iff.2 fun y yfs => _
-  rcases mem_image_iff_bex.1 yfs with ⟨x, xs, fxy⟩
+  rcases mem_image.1 yfs with ⟨x, xs, fxy⟩
```
```suggestion
  rcases yfs with ⟨x, xs, fxy⟩
```

YaelDillies commented on Mathlib/Data/List/Basic.lean:
```diff
@@ -275,7 +275,7 @@ theorem forall_mem_of_forall_mem_cons {p : α → Prop} {a : α} {l : List α} (
 #align list.forall_mem_append List.forall_mem_append
 
-theorem not_exists_mem_nil (p : α → Prop) : ¬∃ x ∈ @nil α, p x :=
+theorem not_exists_nil (p : α → Prop) : ¬∃ x ∈ @nil α, p x :=
```
```suggestion
theorem not_exists_mem_nil (p : α → Prop) : ¬∃ x ∈ @nil α, p x :=
```

grunweg replied:
This lemma already exists (line 75 of that file): as-is, this change doesn't compile

YaelDillies replied:
Can you deduplicate then?

grunweg replied:
AFAICT, this theorem is never used in mathlib: I've removed it (and kept the align, as it points to the other lemma).

YaelDillies commented on Mathlib/Data/Set/Basic.lean:
```diff
@@ -1206,15 +1207,17 @@ theorem forall_insert_of_forall {P : α → Prop} {a : α} {s : Set α} (H : ∀
 /- Porting note: ∃ x ∈ insert a s, P x is parsed as ∃ x, x ∈ insert a s ∧ P x,
  where in Lean3 it was parsed as `∃ x, ∃ (h : x ∈ insert a s), P x` -/
-theorem bex_insert_iff {P : α → Prop} {a : α} {s : Set α} :
+theorem exists_insert_iff {P : α → Prop} {a : α} {s : Set α} :
```
```suggestion
theorem exists_mem_insert {P : α → Prop} {a : α} {s : Set α} :
```

YaelDillies commented on Mathlib/Data/Set/Basic.lean:
```diff
@@ -1206,15 +1207,17 @@ theorem forall_insert_of_forall {P : α → Prop} {a : α} {s : Set α} (H : ∀
+@[deprecated] alias bex_insert_iff := exists_insert_iff -- 2024-03-23
 
-theorem ball_insert_iff {P : α → Prop} {a : α} {s : Set α} :
+theorem forall_insert_iff {P : α → Prop} {a : α} {s : Set α} :
```
```suggestion
theorem forall_mem_insert {P : α → Prop} {a : α} {s : Set α} :
```

YaelDillies commented on Mathlib/GroupTheory/Subgroup/Basic.lean:
```diff
@@ -1403,7 +1403,7 @@ theorem coe_map (f : G →* N) (K : Subgroup G) : (K.map f : Set N) = f '' K :=
 
 @[to_additive (attr := simp)]
 theorem mem_map {f : G →* N} {K : Subgroup G} {y : N} : y ∈ K.map f ↔ ∃ x ∈ K, f x = y := by
```
```suggestion
theorem mem_map {f : G →* N} {K : Subgroup G} {y : N} : y ∈ K.map f ↔ ∃ x ∈ K, f x = y := Iff.rfl
```

YaelDillies commented on Mathlib/GroupTheory/Submonoid/Operations.lean:
```diff
@@ -232,8 +232,7 @@ theorem coe_map (f : F) (S : Submonoid M) : (S.map f : Set N) = f '' S :=
 
 @[to_additive (attr := simp)]
 theorem mem_map {f : F} {S : Submonoid M} {y : N} : y ∈ S.map f ↔ ∃ x ∈ S, f x = y := by
```
```suggestion
theorem mem_map {f : F} {S : Submonoid M} {y : N} : y ∈ S.map f ↔ ∃ x ∈ S, f x = y := Iff.rfl
```
so that it can be used by `dsimp`

grunweg commented:
Updated again.

YaelDillies commented on Mathlib/RingTheory/NonUnitalSubsemiring/Basic.lean:
```diff
@@ -304,8 +304,7 @@ theorem coe_map (f : F) (s : NonUnitalSubsemiring R) : (s.map f : Set S) = f ''
 
 @[simp]
 theorem mem_map {f : F} {s : NonUnitalSubsemiring R} {y : S} : y ∈ s.map f ↔ ∃ x ∈ s, f x = y := by
```
```suggestion
theorem mem_map {f : F} {s : NonUnitalSubsemiring R} {y : S} : y ∈ s.map f ↔ ∃ x ∈ s, f x = y := Iff.rfl
```
so that it can be used by `dsimp`

YaelDillies commented on Mathlib/Analysis/LocallyConvex/WithSeminorms.lean:
```diff
@@ -547,7 +547,7 @@ theorem WithSeminorms.isVonNBounded_iff_seminorm_bounded {s : Set E} (hp : WithS
   · choose r hr h using hi
     have h' : 0 < I.sup' hI r := by
-      rcases hI.bex with ⟨i, hi⟩
+      rcases hI.exists with ⟨i, hi⟩
```
```suggestion
      rcases hI with ⟨i, hi⟩
```

YaelDillies commented on Mathlib/RingTheory/Ideal/Operations.lean:
```diff
@@ -1347,7 +1347,7 @@ theorem subset_union_prime {R : Type u} [CommRing R] {s : Finset ι} {f : ι →
         have : (I : Set R) ≠ ∅ := Set.Nonempty.ne_empty (Set.nonempty_of_mem I.zero_mem)
         exact absurd h this
-      · cases' hsne.bex with i his
+      · cases' hsne.exists with i his
```
```suggestion
      · cases' hsne with i his
```

urkud commented:
Thanks!
bors d=@YaelDillies

YaelDillies commented:
> in Logic.Basic: this has lemmas like `forall_of_ball` and conversely, so it's not obvious how to name them

Can you simply replace `ball` by `forall_mem` in those names?

grunweg commented:
I went over all lemmas in Logic.Basic, replacing "bex" by "exists_mem" and "ball" by "forall_mem". Oof.

There are a couple of occurrences of "bef_def"; be my guest to remove them :-)

YaelDillies commented on Mathlib/Logic/Basic.lean:
```diff
@@ -1060,61 +1069,79 @@ theorem BEx.imp_left (H : ∀ x, p x → q x) : (∃ (x : _) (_ : p x), r x) →
-#align exists_of_bex exists_of_bex
+#align exists_of_bex exists_of_exists_mem
+
+theorem exists_mem_imp : (∃ x h, P x h) → b ↔ ∀ x h, P x h → b := by simp
```
```suggestion
theorem exists₂_imp : (∃ x h, P x h) → b ↔ ∀ x h, P x h → b := by simp
```

grunweg commented:
Do you have an opinion about `{BAll,BEx}.imp_{left,right}` in `Logic.Basic`?
I could rename them to Forall/Exists.imp_{left,right} instead.

YaelDillies commented on Mathlib/Logic/Basic.lean:
```diff
@@ -1060,61 +1069,79 @@ theorem BEx.imp_left (H : ∀ x, p x → q x) : (∃ (x : _) (_ : p x), r x) →
 
-theorem ball_of_forall (h : ∀ x, p x) (x) : p x := h x
-#align ball_of_forall ball_of_forall
+theorem forall_mem_of_forall (h : ∀ x, p x) (x) : p x := h x
+#align ball_of_forall forall_mem_of_forall
```
Actually this is a stupid lemma. Can you revert the rename and deprecate it to `id`?

YaelDillies commented on Mathlib/Logic/Basic.lean:
```diff
@@ -1060,61 +1069,79 @@ theorem BEx.imp_left (H : ∀ x, p x → q x) : (∃ (x : _) (_ : p x), r x) →
 
-theorem forall_of_ball (H : ∀ x, p x) (h : ∀ x, p x → q x) (x) : q x := h x <| H x
-#align forall_of_ball forall_of_ball
+theorem forall_of_forall_mem (H : ∀ x, p x) (h : ∀ x, p x → q x) (x) : q x := h x <| H x
+#align forall_of_ball forall_of_forall_mem
```
Can you revert the rename and deprecate to `forall_imp`?

YaelDillies commented on Mathlib/Logic/Basic.lean:
```diff
@@ -1060,61 +1069,79 @@ theorem BEx.imp_left (H : ∀ x, p x → q x) : (∃ (x : _) (_ : p x), r x) →
+#align not_bex not_exists_mem
 
-theorem not_ball_of_bex_not : (∃ x h, ¬P x h) → ¬∀ x h, P x h
+theorem not_forall_mem_of_exists_mem_not : (∃ x h, ¬P x h) → ¬∀ x h, P x h
```
```suggestion
theorem not_forall₂_of_exists₂_not : (∃ x h, ¬P x h) → ¬∀ x h, P x h
```

YaelDillies commented on Mathlib/Logic/Basic.lean:
```diff
@@ -1060,61 +1069,79 @@ theorem BEx.imp_left (H : ∀ x, p x → q x) : (∃ (x : _) (_ : p x), r x) →
 
 -- See Note [decidable namespace]
-protected theorem Decidable.not_ball [Decidable (∃ x h, ¬P x h)] [∀ x h, Decidable (P x h)] :
+protected theorem Decidable.not_forall_mem [Decidable (∃ x h, ¬P x h)] [∀ x h, Decidable (P x h)] :
```
```suggestion
protected theorem Decidable.not_forall₂ [Decidable (∃ x h, ¬P x h)] [∀ x h, Decidable (P x h)] :
```

YaelDillies commented on Mathlib/Logic/Basic.lean:
```diff
@@ -1060,61 +1069,79 @@ theorem BEx.imp_left (H : ∀ x, p x → q x) : (∃ (x : _) (_ : p x), r x) →
 
-theorem not_ball : (¬∀ x h, P x h) ↔ ∃ x h, ¬P x h := Decidable.not_ball
-#align not_ball not_ball
+theorem not_forall_mem : (¬∀ x h, P x h) ↔ ∃ x h, ¬P x h := Decidable.not_forall_mem
+#align not_ball not_forall_mem
```
```suggestion
theorem not_forall₂ : (¬∀ x h, P x h) ↔ ∃ x h, ¬P x h := Decidable.not_forall_mem
#align not_ball not_forall₂
```

YaelDillies commented on Mathlib/Logic/Basic.lean:
```diff
@@ -1060,61 +1069,79 @@ theorem BEx.imp_left (H : ∀ x, p x → q x) : (∃ (x : _) (_ : p x), r x) →
+#align not_ball not_forall_mem
 
-theorem ball_true_iff (p : α → Prop) : (∀ x, p x → True) ↔ True :=
+theorem forall_mem_true_iff (p : α → Prop) : (∀ x, p x → True) ↔ True :=
```
```suggestion
theorem forall₂_true_iff (p : α → Sort*) : (∀ x, p x → True) ↔ True :=
```

grunweg replied:
As this lemma already existed, I deduplicated this one instead.

YaelDillies commented on Mathlib/Logic/Basic.lean:
```diff
@@ -1060,61 +1069,79 @@ theorem BEx.imp_left (H : ∀ x, p x → q x) : (∃ (x : _) (_ : p x), r x) →
+#align bex_or_distrib exists_mem_or
 
-theorem ball_or_left : (∀ x, p x ∨ q x → r x) ↔ (∀ x, p x → r x) ∧ ∀ x, q x → r x :=
+theorem forall_mem_or_left : (∀ x, p x ∨ q x → r x) ↔ (∀ x, p x → r x) ∧ ∀ x, q x → r x :=
```
```suggestion
theorem forall₂_or_left : (∀ x, p x ∨ q x → r x) ↔ (∀ x, p x → r x) ∧ ∀ x, q x → r x :=
```
etc

grunweg replied:
Done: should `forall_mem_congr` be left as-is? Or perhaps, my real question is: what do we use "forall₂" for?

YaelDillies replied:
`∀ x ∈ s,` is `forall_mem`. `∀ x y,` is `forall₂`

grunweg replied:
:bulb: Thank you, not it makes sense!

grunweg commented:
> Do you have an opinion about `{BAll,BEx}.imp_{left,right}` in `Logic.Basic`? I could rename them to Forall/Exists.imp_{left,right} instead.

There are also BEx.intro and BEx.elim remaining.

grunweg commented:
Thanks for the careful comments again; I have addressed them.

YaelDillies commented on Mathlib/Logic/Basic.lean:
```diff
@@ -1032,17 +1036,22 @@ theorem BEx.intro (a : α) (h₁ : p a) (h₂ : P a h₁) : ∃ (x : _) (h : p x
 #align bex.intro BEx.intro
 
-theorem ball_congr (H : ∀ x h, P x h ↔ Q x h) : (∀ x h, P x h) ↔ ∀ x h, Q x h :=
+theorem forall_mem_congr (H : ∀ x h, P x h ↔ Q x h) : (∀ x h, P x h) ↔ ∀ x h, Q x h :=
```
```suggestion
theorem forall₂_congr (H : ∀ x h, P x h ↔ Q x h) : (∀ x h, P x h) ↔ ∀ x h, Q x h :=
```

grunweg replied:
This already exists, with a different signature: https://leanprover-community.github.io/mathlib4_docs/Init/PropLemmas.html#forall%E2%82%82_congr
Replacing them wasn't obvious to me at this hour.

YaelDillies replied:
Done

YaelDillies commented on Mathlib/Logic/Basic.lean:
```diff
@@ -1032,17 +1036,22 @@ theorem BEx.intro (a : α) (h₁ : p a) (h₂ : P a h₁) : ∃ (x : _) (h : p x
+#align ball_congr forall_mem_congr
 
-theorem bex_congr (H : ∀ x h, P x h ↔ Q x h) : (∃ x h, P x h) ↔ ∃ x h, Q x h :=
+theorem exists_mem_congr (H : ∀ x h, P x h ↔ Q x h) : (∃ x h, P x h) ↔ ∃ x h, Q x h :=
```
```suggestion
theorem exists₂_congr (H : ∀ x h, P x h ↔ Q x h) : (∃ x h, P x h) ↔ ∃ x h, Q x h :=
```

grunweg replied:
This already exists, with a different signature: https://leanprover-community.github.io/mathlib4_docs/Init/PropLemmas.html#exists%E2%82%82_congr
Replacing them wasn't obvious to me at this hour.

YaelDillies replied:
I know 😉 The lemma here is a special case of the existing `exists₂_congr`

YaelDillies commented on Mathlib/Logic/Basic.lean:
```diff
@@ -1032,17 +1036,22 @@ theorem BEx.intro (a : α) (h₁ : p a) (h₂ : P a h₁) : ∃ (x : _) (h : p x
+#align bex_congr exists_mem_congr
 
-theorem bex_eq_left {a : α} : (∃ (x : _) (_ : x = a), p x) ↔ p a := by
+theorem exists_mem_eq_left {a : α} : (∃ (x : _) (_ : x = a), p x) ↔ p a := by
```
```suggestion
theorem exists_eq_left {a : α} : (∃ (x : _) (_ : x = a), p x) ↔ p a := by
```

grunweg replied:
Same here, this already exists at: https://leanprover-community.github.io/mathlib4_docs/Init/PropLemmas.html#exists_eq_left

YaelDillies replied:
This one I think you can just deprecate since it's quite convoluted and simplifiable by simp already

YaelDillies commented on Mathlib/Logic/Basic.lean:
```diff
@@ -1060,61 +1069,77 @@ theorem BEx.imp_left (H : ∀ x, p x → q x) : (∃ (x : _) (_ : p x), r x) →
+#align ball_and_distrib forall₂_and
 
-theorem bex_or : (∃ x h, P x h ∨ Q x h) ↔ (∃ x h, P x h) ∨ ∃ x h, Q x h :=
+theorem exists_mem_or : (∃ x h, P x h ∨ Q x h) ↔ (∃ x h, P x h) ∨ ∃ x h, Q x h :=
```
```suggestion
theorem exists₂_or : (∃ x h, P x h ∨ Q x h) ↔ (∃ x h, P x h) ∨ ∃ x h, Q x h :=
```

YaelDillies commented on Mathlib/Logic/Basic.lean:
```diff
@@ -1060,61 +1069,77 @@ theorem BEx.imp_left (H : ∀ x, p x → q x) : (∃ (x : _) (_ : p x), r x) →
+#align ball_or_left_distrib forall₂_or_left
 
-theorem bex_or_left :
+theorem exists_mem_or_left :
```
```suggestion
theorem exists_or_left :
```

grunweg commented:
Feel free to do any of the remaining three lemmas yourself, if it's faster this way.

YaelDillies commented:
Must be gone for the night, but I will if you haven't done them in the next 8-9 hours.

grunweg commented:
@YaelDillies I just rebased this and fixed the one easy lemma. Do you have energy left to fix the two remaining lemmas?

In any case: thanks for all your help on this PR!

YaelDillies reviewed (approved):
Now done. LGTM

maintainer merge

jcommelin reviewed (approved):
Thanks :tada:

bors merge
